roadmap/implementers-guide/src/runtime/scheduler.md

```diff
@@ -1,3 +1,269 @@
+         self.parachain_provider.pop_assignment(chain_id),
+     }
+   }
+}
+// What core ranges belong to which provider will be configurable and can change
+// at session boundaries.
+```
+
+The providers need to be able to handle some cores moving faster than others.
+For example, if availability is slower on one core for some reason, then the
+scheduler will call `pop_assignment_for_core` less on that core ... it will
+impose some back pressure. The provider needs to be able to handle this.
+
+Open questions:
+
+With this interface we will charge parathreads fully once
+`pop_assignment_for_core` got called, no matter what happens next. That is even
+on error. This adds to simplicity, great if acceptable ... I think at least for
+a first run it is. This is one of those things that can be improved, once it is
+actually causing problems. -> Especially with async backing it is actually quite
+unlikely that things fail.
+
+
+### Interface from the scheduler to the node size (collators/validators)
+
+We have been a bit fuzzy with the claim queue here, time to make it more
+concrete. The claim queue is used internally by the scheduler to be able to
+satisfy a lookahead on assignments. It could look something like this:
+
+```rust
+/// A claim queue for a single core:
+struct ClaimQueue {
+  /// Actual queue:
+  queue: Vec<ParaId>
+  /// A claim previously in queue directly, that is now pending availability.
+  ///
+  /// It is counting towards the queue size, in the sense that we only request
+  /// another assignment once the core is freed again.
+  pending_availability: PendingAvailabilityClam,
+
+}
+
+struct PendingAvailabilityClaim {
+  /// The para that is currently occupying the core.
+  para: ParaId,
+  /// The position in the queue the claim had, when it got backed.
+  /// Used to determine validator era points.
+  pos_in_queue: usize,
+}
+```
+
+We can then provide a runtime API to the node, for retrieving the current claim
+queue for each core. What we will expose here is all the claims, except the head
+if it is already backed. Therefore we would expose actual viable claims for
+upcoming blocks, once the core is free again.
+
+The above defined `ClaimQueue` already provides the means to allow later claims
+to be backed before claims further up the list, as described
+[here](https://github.com/paritytech/polkadot/issues/5492#issuecomment-1362941241).
+By tracking the position the para had in the list, also adjusting the rewards
+accordingly should be straight forward.
+
+
+# Parathreads - other areas:
+
+## Cumulus Integration
+
+### Key Management
+
+Collators need to be able to bid securely in auctions for getting a slot. Idea:
+Have restricted (proxy) accounts which are restricted to bidding on parathread
+auctions for a specific `ParaId`. This can further be restricted to just a bid
+per day/per hour/.... whatever is suitable for the parathread and greatly
+reduces possible damage if keys are leaked/lost.
```

Comment:
These should be session keys. An account on the relay chain could set its current collator session key. We could then let collators send unsigned transactions for bidding. If some locked funds on the account setting the collator key isn't enough, we could always deduct fees for the unsigned transaction from the account. The payment for the block would then always come from the account that set the keys. No one could then do that much bad with session keys if they are leaked, besides trying to bid for slots and burning the money in this way.

Reply:
https://github.com/orgs/paritytech/projects/67/views/1?pane=issue&itemId=19145863 Have you seen this? I think a different idea is to have collators store a "hot" key which is a proxy that is only allowed to issue bids.

Collators are currently not registered on the relay chain, and any account can issue a bid "on behalf of" any collator. There is no real harm in buying blockspace for arbitrary collators, because the account paying for the blockspace is the one on the hook for the bid and the collator would receive any block rewards / transaction fees.

Reply:
I think we should support both. Giving the people more freedom on how to implement it. Either a collator is "pre-registered" and then uses an unsigned transaction to do the bidding on its own or we have some account that sends some `BidFor { collator: Key, bid: Balance}`. Internally both then map to the same function `bid(collator: Key, bid: Balance)`

Reply:
It is usually better to have one way of doing things than 2. What benefit is there for collators to pre-register?

Reply:
When you pre-register a collator public key you wouldn't need any private key stored on your collator that has any funds in the account on the relay chain.

Reply:
There is nothing you can do with a special "registered collator account" that you couldn't do for any account. It seems like a moot point.

Reply:
No matter what collators will need to be able to automatically send transactions that pay for on-demand blockspace. It is definitely a security issue that collators need to worry about and there is no way around it. Proxy accounts limit the attack surface to only paying for blockspace, at least.

Reply:
We could have a separate pallet to even configure the account to only be allowed to bid on a specific para.

Reply:
User flow:
  1. Fund bidder account A
  2. Set A's `BidParachain` in the `OnDemandBlockspace` pallet to a specific ParaId (optional)
  3. Set A's `BidCollator` in the `OnDemandBlockspace` pallet to a specific CollatorId (optional)
  4. Register `OnDemandBid` proxy P for A
  5. Give collator node the private key for proxy P
  6. collator sends `OnDemandBlockspace::bid` transactions whenever it needs to
  7. `OnDemandBlockspace::bid` checks optional para-id and collator ID.

This is pure FRAME without special-casing.

Reply:
You would also need to have enough funds to pay transaction fees and constantly ensure that the account isn't running below ED. I mean you are right that these aren't hard technical points or change the fact that an attacker could still attack by draining the funds for bidding, but I think it makes the usage easier. But this isn't such an important topic and we could still later open the route for unsigned transactions. This isn't really any point that will change the implementation fundamentalally.

Reply:
After a little research I like the proxy approach better as well (for now), because signed extrinsics come with a bunch of security features by default, which would need to be re-implemented for the unsigned case. Might not be a big deal, but I generally like not re-implementing the wheel, especially when it comes to security relevant matters.



> You would also need to have enough funds to pay transaction fees and constantly ensure that the account isn't running below ED.

Wouldn't the same hold true for unsigned extrinsics signed by the `CollatorId`? Orders would still not be free, but will be charged to some account (coupled to the `CollatorId`).

Reply:
> Wouldn't the same hold true for unsigned extrinsics signed by the `CollatorId`? Orders would still not be free, but will be charged to some account (coupled to the `CollatorId`).

By orders you mean the bids? When you are using the proxy approach, you will need two accounts with funds in it. The account that is used to bid (which has all the funds in it to pay the bids) and the proxy account that controls the bidding (this is the account stored on the collator and requires funds to pay the tx fees). The collator will send transactions using the proxy account to bid with the bidding account. When we would have an unsigned extrinsic you would not need any funds to send the unsigned extrinsic as unsigned extrinsic by design don't pay any fees by default. Then you only need to have funds in the bidding accounts to pay the bids.

Reply:
Got it. Yeah, I already noticed that proxy accounts are a bit of a pain from the user interface perspective. But using an unsigned extrinsic to avoid the bad user interface of proxy accounts seems a bit backwards - would it be hard to improve proxy accounts to benefit all use cases and not just parathreads? I don't see a fundamental reason why fees for transactions of a proxy account should not be paid by the proxied account - in fact we would need that for using the proxy account abstraction here anyway, because I am aiming on customizing fees to pay for the order as well, if that would be charged to the proxy account instead of the proxied account, then this would become rather pointless.

Looks like looking into customizing fees might be pointless after all. 

> By orders you mean the bids? 

Yes orders are bids, because in our current model they are actually not bids anymore, but rather orders - which are either valid or invalid, but other than that the same.

Reply:
> would it be hard to improve proxy accounts to benefit all use cases and not just parathreads?

https://github.com/paritytech/substrate/issues/7054


> because I am aiming on customizing fees to pay for the order as well

Fees are currently based on the weight and you can not that easily change this. However, with some custom signed extension it should maybe be possible.